# Home Credit Risk Analysis

## Dataset: application_train.csv

### Objetivo 

Analizar la estructura, calidad y relaciones del dataset principal de solicitudes de crédito para comprender su papel dentro de la arquitectura Lakehouse y del modelo dimensional. 

---

### Contexto del negocio 

Home credit es una entidad financiera que busca predecir el riesgo de incumplimiento de pago de sus clientes. 

Este dataset contiene la información principal de las solicitudes de crédito y será utilizado como punto de partida para la construcción de las capas: 

- Bronze 
- Silver 
- Intermediate 
- Gold 
- Diamond 

Dentro de la arquitectura híbrida ETL/ELT.

# 1. Importación de librerías 

Se cargan las librerías necesarias para el análisis exploratorio de datos. 

In [1]:
import pandas as pd 
import numpy as np 

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 200)

# 2. Carga del Dataset 

Se carga el archivo `application_train.csv`, considerado el dataset principal del dominio de negocio. 

Este dataset representa las solicitudes de crédito realizadas por los clientes. 

In [2]:
df = pd.read_csv("../../../data/raw/homeCredit/application_train.csv")

print(f"Filas: {df.shape[0]}")
print(f"Columnas: {df.shape[1]}")

Filas: 307511
Columnas: 122


### Observaciones iniciales 

El dataset contiene **307.511 solicitudes de crédito** y **122 variables**. 

La cantidad de atributos sugiere que la información proviene de múltiples dominios de negocio: 

- Información demográfica 
- Información laboral 
- Información financiera 
- Información patrimonial 
- Información crediticia 
- Información histórica de comportamiento 

Debido a su natiraleza central, se espera que este dataset actúe cómo la principal tabla de hechos del modelo dimensiona: 

**FACT_CREDIT_APPLICATION**

# 3. Vista preliminar de los datos 

Se inspeccionan los primeros registros para comprender la estructura general y el tipo de información almacenada.

In [3]:
df.head()

,SK_ID_CURR,TARGET,NAME_CONTRACT_TYPE,CODE_GENDER,FLAG_OWN_CAR,FLAG_OWN_REALTY,CNT_CHILDREN,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY,AMT_GOODS_PRICE,NAME_TYPE_SUITE,NAME_INCOME_TYPE,NAME_EDUCATION_TYPE,NAME_FAMILY_STATUS,NAME_HOUSING_TYPE,REGION_POPULATION_RELATIVE,DAYS_BIRTH,DAYS_EMPLOYED,DAYS_REGISTRATION,DAYS_ID_PUBLISH,OWN_CAR_AGE,FLAG_MOBIL,FLAG_EMP_PHONE,FLAG_WORK_PHONE,FLAG_CONT_MOBILE,FLAG_PHONE,FLAG_EMAIL,OCCUPATION_TYPE,CNT_FAM_MEMBERS,REGION_RATING_CLIENT,REGION_RATING_CLIENT_W_CITY,WEEKDAY_APPR_PROCESS_START,HOUR_APPR_PROCESS_START,REG_REGION_NOT_LIVE_REGION,REG_REGION_NOT_WORK_REGION,LIVE_REGION_NOT_WORK_REGION,REG_CITY_NOT_LIVE_CITY,REG_CITY_NOT_WORK_CITY,LIVE_CITY_NOT_WORK_CITY,ORGANIZATION_TYPE,EXT_SOURCE_1,EXT_SOURCE_2,EXT_SOURCE_3,APARTMENTS_AVG,BASEMENTAREA_AVG,YEARS_BEGINEXPLUATATION_AVG,YEARS_BUILD_AVG,COMMONAREA_AVG,ELEVATORS_AVG,ENTRANCES_AVG,FLOORSMAX_AVG,FLOORSMIN_AVG,LANDAREA_AVG,LIVINGAPARTMENTS_AVG,LIVINGAREA_AVG,NONLIVINGAPARTMENTS_AVG,NONLIVINGAREA_AVG,APARTMENTS_MODE,BASEMENTAREA_MODE,YEARS_BEGINEXPLUATATION_MODE,YEARS_BUILD_MODE,COMMONAREA_MODE,ELEVATORS_MODE,ENTRANCES_MODE,FLOORSMAX_MODE,FLOORSMIN_MODE,LANDAREA_MODE,LIVINGAPARTMENTS_MODE,LIVINGAREA_MODE,NONLIVINGAPARTMENTS_MODE,NONLIVINGAREA_MODE,APARTMENTS_MEDI,BASEMENTAREA_MEDI,YEARS_BEGINEXPLUATATION_MEDI,YEARS_BUILD_MEDI,COMMONAREA_MEDI,ELEVATORS_MEDI,ENTRANCES_MEDI,FLOORSMAX_MEDI,FLOORSMIN_MEDI,LANDAREA_MEDI,LIVINGAPARTMENTS_MEDI,LIVINGAREA_MEDI,NONLIVINGAPARTMENTS_MEDI,NONLIVINGAREA_MEDI,FONDKAPREMONT_MODE,HOUSETYPE_MODE,TOTALAREA_MODE,WALLSMATERIAL_MODE,EMERGENCYSTATE_MODE,OBS_30_CNT_SOCIAL_CIRCLE,DEF_30_CNT_SOCIAL_CIRCLE,OBS_60_CNT_SOCIAL_CIRCLE,DEF_60_CNT_SOCIAL_CIRCLE,DAYS_LAST_PHONE_CHANGE,FLAG_DOCUMENT_2,FLAG_DOCUMENT_3,FLAG_DOCUMENT_4,FLAG_DOCUMENT_5,FLAG_DOCUMENT_6,FLAG_DOCUMENT_7,FLAG_DOCUMENT_8,FLAG_DOCUMENT_9,FLAG_DOCUMENT_10,FLAG_DOCUMENT_11,FLAG_DOCUMENT_12,FLAG_DOCUMENT_13,FLAG_DOCUMENT_14,FLAG_DOCUMENT_15,FLAG_DOCUMENT_16,FLAG_DOCUMENT_17,FLAG_DOCUMENT_18,FLAG_DOCUMENT_19,FLAG_DOCUMENT_20,FLAG_DOCUMENT_21,AMT_REQ_CREDIT_BUREAU_HOUR,AMT_REQ_CREDIT_BUREAU_DAY,AMT_REQ_CREDIT_BUREAU_WEEK,AMT_REQ_CREDIT_BUREAU_MON,AMT_REQ_CREDIT_BUREAU_QRT,AMT_REQ_CREDIT_BUREAU_YEAR
0,100002,1,Cash loans,M,N,Y,0,202500.0,406597.5,24700.5,351000.0,Unaccompanied,Working,Secondary / secondary special,Single / not married,House / apartment,0.018801,-9461,-637,-3648.0,-2120,NaN,1,1,0,1,1,0,Laborers,1.0,2,2,WEDNESDAY,10,0,0,0,0,0,0,Business Entity Type 3,0.083037,0.262949,0.139376,0.0247,0.0369,0.9722,0.6192,0.0143,0.00,0.0690,0.0833,0.1250,0.0369,0.0202,0.0190,0.0000,0.0000,0.0252,0.0383,0.9722,0.6341,0.0144,0.0000,0.0690,0.0833,0.1250,0.0377,0.022,0.0198,0.0,0.0,0.0250,0.0369,0.9722,0.6243,0.0144,0.00,0.0690,0.0833,0.1250,0.0375,0.0205,0.0193,0.0000,0.00,reg oper account,block of flats,0.0149,"Stone, brick",No,2.0,2.0,2.0,2.0,-1134.0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.0,0.0,0.0,0.0,0.0,1.0
1,100003,0,Cash loans,F,N,N,0,270000.0,1293502.5,35698.5,1129500.0,Family,State servant,Higher education,Married,House / apartment,0.003541,-16765,-1188,-1186.0,-291,NaN,1,1,0,1,1,0,Core staff,2.0,1,1,MONDAY,11,0,0,0,0,0,0,School,0.311267,0.622246,NaN,0.0959,0.0529,0.9851,0.7960,0.0605,0.08,0.0345,0.2917,0.3333,0.0130,0.0773,0.0549,0.0039,0.0098,0.0924,0.0538,0.9851,0.8040,0.0497,0.0806,0.0345,0.2917,0.3333,0.0128,0.079,0.0554,0.0,0.0,0.0968,0.0529,0.9851,0.7987,0.0608,0.08,0.0345,0.2917,0.3333,0.0132,0.0787,0.0558,0.0039,0.01,reg oper account,block of flats,0.0714,Block,No,1.0,0.0,1.0,0.0,-828.0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0
2,100004,0,Revolving loans,M,Y,Y,0,67500.0,135000.0,6750.0,135000.0,Unaccompanied,Working,Secondary / secondary special,Single / not married,House / apartment,0.010032,-19046,-225,-4260.0,-2531,26.0,1,1,1,1,1,0,Laborers,1.0,2,2,MONDAY,9,0,0,0,0,0,0,Government,NaN,0.555912,0.729567,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,

# 4. Estructura General 

Se analiza la composición del dataset: 

- Cantidad de columnas 
- Tipos de datos 
- Presencia de valores nulos 
- Distribución general de atributos 

In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 307511 entries, 0 to 307510
Columns: 122 entries, SK_ID_CURR to AMT_REQ_CREDIT_BUREAU_YEAR
dtypes: float64(65), int64(41), str(16)
memory usage: 325.2 MB


# 5. Identificación de la Clave Primaria 

Se evalúa si la columna `SK_ID_CURR` puede actuar como identificador único de cada solicitud. 

Esta validación es fundamental para: 

- Integridad de datos 
- Relaciones entre datasets 
- Diseño dimensional 
- Procesos de deduplicación 

In [5]:
total_rows = len(df)

unique_ids = df["SK_ID_CURR"].nunique()

duplicates = df["SK_ID_CURR"].duplicated().sum()

print("Resgistros: ", total_rows)
print("IDs únicos: ", unique_ids)
print("Duplicados: ", duplicates)

Resgistros:  307511
IDs únicos:  307511
Duplicados:  0


## Resultados 

| Métricas | Valor |
|----------|-------|
|Total registros | 307,511 |
|IDs únicos | 307,511 |
|IDs duplicados | 0 |

### Conclusión 

La columna SK_ID_CURR identifica de manera única cada registro. 

Por lo tanto: 

- SK_ID_CURR será considerada la clave primaria. 
- No se detectaron duplicados. 
- Cada fila representa una única solicitud de crédito. 

### Implicaciones 

FACT_CREDIT_APPLICACTION

# 6. Análisis de Calidad de Datos 

Se identifican columnas con información faltante. 

Este análisis permitirá posteriormente definir: 

- Reglas de calidad 
- Estrategias de imputación 
- Transformaciones Silver 

In [11]:
null_percentage = (
    df.isnull()
    .mean()
    .mul(100)
    .sort_values(ascending=False)
)

null_percentage.head(20)

COMMONAREA_AVG              69.872297
COMMONAREA_MODE             69.872297
COMMONAREA_MEDI             69.872297
NONLIVINGAPARTMENTS_MEDI    69.432963
NONLIVINGAPARTMENTS_MODE    69.432963
NONLIVINGAPARTMENTS_AVG     69.432963
FONDKAPREMONT_MODE          68.386172
LIVINGAPARTMENTS_AVG        68.354953
LIVINGAPARTMENTS_MEDI       68.354953
LIVINGAPARTMENTS_MODE       68.354953
FLOORSMIN_MODE              67.848630
FLOORSMIN_AVG               67.848630
FLOORSMIN_MEDI              67.848630
YEARS_BUILD_AVG             66.497784
YEARS_BUILD_MODE            66.497784
YEARS_BUILD_MEDI            66.497784
OWN_CAR_AGE                 65.990810
LANDAREA_MEDI               59.376738
LANDAREA_AVG                59.376738
LANDAREA_MODE               59.376738
dtype: float64

## Principales columnas afectadas 

Las variables con mayor porcentaje de valores faltantes pertenecen principalmente a información inmobiliaria. 

### Conclusión 

Estas columnas deberán ser evaluadas durante la construcción de la capa Silver mediante reglas de calidad e imputación. 

# 7. Detección de Registros Duplicados 

Se verifica la existencia de registros completamente duplicados dentro del dataset. 

In [7]:
df.duplicated().sum()

np.int64(0)

# 8. Varibale Objetivo 

La variable TARGET representa el comportamiento final del crédito. 

Valores: 

- 0 -> CLiente sin incumplimineto 
- 1 -> Cliente con incumplimiento 

Este atributo será utilizado posteriormente para análisis predictivos. 

In [8]:
df["TARGET"].value_counts()

TARGET
0    282686
1     24825
Name: count, dtype: int64

In [9]:
(
    df["TARGET"]
    .value_counts(normalize=True)
    .mul(100)
)

TARGET
0    91.927118
1     8.072882
Name: proportion, dtype: float64

TARGET reoresenta el resultado final del crédito. 

| Valor | Significado |
|-------|-------------|
|0| Cliente sin incumplimiento|
|1| Cliente con incumplimineto| 

### Observación 

Existe un desbalance importante entre clases, siendo mucho más frecuente el no incumplimiento. 

# 9. Variables Categóricas 

Se identifican las columnas categóricas presentes en el dataset. 

Estas variables son candidatas a: 

- Dimensiones 
- Catálogos de referencia 
- Transformaciones de codificación 

In [10]:
categorical_columns = df.select_dtypes(include="object").columns

categorical_columns.tolist()

C:\Users\ferna\AppData\Local\Temp\ipykernel_14912\859468477.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_columns = df.select_dtypes(include="object").columns


['NAME_CONTRACT_TYPE',
 'CODE_GENDER',
 'FLAG_OWN_CAR',
 'FLAG_OWN_REALTY',
 'NAME_TYPE_SUITE',
 'NAME_INCOME_TYPE',
 'NAME_EDUCATION_TYPE',
 'NAME_FAMILY_STATUS',
 'NAME_HOUSING_TYPE',
 'OCCUPATION_TYPE',
 'WEEKDAY_APPR_PROCESS_START',
 'ORGANIZATION_TYPE',
 'FONDKAPREMONT_MODE',
 'HOUSETYPE_MODE',
 'WALLSMATERIAL_MODE',
 'EMERGENCYSTATE_MODE']

# 10. Conclusiones Técnicas 


## Resumen del Dataset 

| Métrica    | Valor      |
|------------|------------|
| Filas      |   307.511  |
| Columnas   |     122    |
| PK         | SK_ID_CURR |
| Duplicados |      0     |
| Target     |   TARGET   |

---

#### Interpretación 

El dataset presenta una alta dimensionalidad, con 122 atributos por registro. 

Cada fila representa una solicitud de crédito realizada por un cliente dentro del programa Home Credit. 

Este dataset constituye la fuente principal de información para el análisis de riesgo crediticio y será utilizado como entidad central dentro del modelo dimensional. 

---

## Calidad de datos 

Se identifican múltiples columnas con porcentajes elevados de valores faltantes, principalmente relacionadas con características inmobiliarias y patrimoniales de los clientes. 

Las columnas más afectadas presentan porcentajes de nulidad superiores al 60%. 

Estas observaciones serán consideradas durante el diseño de las reglas de calidad de la capa Silver. 

---

## Granularidad 

Cada registro representa una única solicitud de crédito asociada a un cliente. 

La columna `SK-ID_CURR` identifica de manera única cada observación. 

--- 

## Relaciones identificadas 

Este dataset actúa como entidad raíz dentro del ecosistema Home Credit y se relaciona con: 

- bureau
- bureau_balance
- previus_application
- installments_payments
- POS_CASH_balance
- credit_card_balance

--- 

## Rol dentro del modelo dimensional 

Entidad principal: 

FACT_CREDIT_APPLICACTION

Desde esta tabla se derivarán las relaciones hacía las dimensiones de cliente, crédito, comportamiento financiero y pagos. 

--- 

## Flujo Lakehouse 

applicaction_train.csv 

↓ Bronze (ingesta cruda)

↓ Silver (calidad y normalización)

↓ Intermediate (integración y enriquecimiento)

↓ Gold (analítica de negocio)

↓ Diamond (convergencia corporativa)